File: FoodDetectionApp/SyncManager.swift
import Foundation
import Combine
import UIKit

/// Syncs local food logs with the server via NetworkService.
/// Uses `/api/sync/push` to upload unsynced logs and `/api/sync/pull` to fetch remote changes.
class SyncManager: ObservableObject {
    static let shared = SyncManager()

    @Published var isSyncing = false
    @Published var lastSyncDate: Date?

    private var cancellables = Set<AnyCancellable>()
    private let isoFormatter = ISO8601DateFormatter()

    private init() {
        if let timestamp = UserDefaults.standard.object(forKey: "lastSyncTimestamp") as? Double {
            lastSyncDate = Date(timeIntervalSince1970: timestamp)
        }

        NotificationCenter.default.publisher(for: UIApplication.willEnterForegroundNotification)
            .sink { [weak self] _ in
                self?.syncIfNeeded()
            }
            .store(in: &cancellables)
    }

    // MARK: - Public

    func syncIfNeeded() {
        guard AuthManager.shared.isSignedIn,
              NetworkMonitor.shared.isConnected,
              !isSyncing else {
            return
        }

        Task { @MainActor in
            isSyncing = true
            defer { isSyncing = false }

            do {
                try await pushLocalLogs()
                try await pullRemoteLogs()

                let now = Date()
                lastSyncDate = now
                UserDefaults.standard.set(now.timeIntervalSince1970, forKey: "lastSyncTimestamp")
            } catch {
                print("[SyncManager] sync failed: \(error.localizedDescription)")
            }
        }
    }

    // MARK: - Push

    private func pushLocalLogs() async throws {
        let allLogs = NutritionManager.shared.logs
        let logsToSync = lastSyncDate.map { since in
            allLogs.filter { $0.time > since }
        } ?? allLogs
        guard !logsToSync.isEmpty else { return }

        let formatter = isoFormatter
        let payload = SyncPushRequest(logs: logsToSync.map { log in
            SyncFoodLogDTO(
                clientId: log.id.uuidString,
                dishName: log.food,
                calories: log.calories,
                proteinG: log.protein,
                carbsG: log.carbs,
                fatsG: log.fats,
                micronutrients: log.micros,
                healthierRecipe: log.recipe,
                loggedAt: formatter.string(from: log.time)
            )
        })

        let _: SyncPushResponse = try await NetworkService.shared.post("/api/sync/push", body: payload)
    }

    // MARK: - Pull

    private func pullRemoteLogs() async throws {
        var queryItems: [URLQueryItem]? = nil
        if let last = lastSyncDate {
            queryItems = [URLQueryItem(name: "since", value: isoFormatter.string(from: last))]
        }

        let response: SyncPullResponse = try await NetworkService.shared.get("/api/sync/pull", queryItems: queryItems)

        guard !response.data.isEmpty else { return }

        let formatter = isoFormatter
        let existingIds = Set(NutritionManager.shared.logs.map { $0.id.uuidString })
        let newLogs = response.data.compactMap { remote -> NutritionManager.FoodLog? in
            guard !existingIds.contains(remote.id) else { return nil }
            return NutritionManager.FoodLog(
                id: UUID(uuidString: remote.id) ?? UUID(),
                food: remote.dishName,
                calories: Int(remote.calories ?? 0),
                protein: Int(remote.proteinG ?? 0),
                carbs: Int(remote.carbsG ?? 0),
                fats: Int(remote.fatsG ?? 0),
                micros: remote.micronutrients,
                recipe: remote.healthierRecipe,
                time: formatter.date(from: remote.loggedAt) ?? Date()
            )
        }

        if !newLogs.isEmpty {
            await MainActor.run {
                NutritionManager.shared.logs.append(contentsOf: newLogs)
                NutritionManager.shared.saveLogs()
            }
        }
    }
}

// MARK: - DTOs

private struct SyncPushRequest: Encodable {
    let logs: [SyncFoodLogDTO]
}

private struct SyncFoodLogDTO: Encodable {
    let clientId: String
    let dishName: String
    let calories: Int
    let proteinG: Int
    let carbsG: Int
    let fatsG: Int
    let micronutrients: [String: String]?
    let healthierRecipe: String?
    let loggedAt: String
}

private struct SyncPushResponse: Decodable {
    let success: Bool
    let data: SyncPushData
}

private struct SyncPushData: Decodable {
    let upserted: Int
}

private struct SyncPullResponse: Decodable {
    let success: Bool
    let data: [RemoteFoodLog]
}

private struct RemoteFoodLog: Decodable {
    let id: String
    let dishName: String
    let calories: Double?
    let proteinG: Double?
    let carbsG: Double?
    let fatsG: Double?
    let micronutrients: [String: String]?
    let healthierRecipe: String?
    let loggedAt: String
}

File: android/app/src/main/java/com/foodsense/android/services/SyncManager.kt
package com.foodsense.android.services

import android.content.Context
import android.util.Log
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import com.foodsense.android.data.FoodLog
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.buildJsonArray
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put
import java.time.Instant
import java.time.format.DateTimeFormatter

class SyncManager(
    private val context: Context,
    private val networkService: NetworkService,
    private val nutritionManager: NutritionManager,
) {
    var isSyncing by mutableStateOf(false)
        private set
    var lastSyncTimeMillis by mutableStateOf(0L)
        private set

    private val scope = CoroutineScope(Dispatchers.IO)
    private val prefs = context.getSharedPreferences("sync_prefs", Context.MODE_PRIVATE)
    private val json = Json {
        ignoreUnknownKeys = true
        isLenient = true
        explicitNulls = false
    }

    init {
        lastSyncTimeMillis = prefs.getLong("lastSyncTimeMillis", 0L)
    }

    fun syncIfNeeded() {
        if (isSyncing) return

        scope.launch {
            isSyncing = true
            try {
                pushLocalLogs()
                pullRemoteLogs()

                val now = System.currentTimeMillis()
                lastSyncTimeMillis = now
                prefs.edit().putLong("lastSyncTimeMillis", now).apply()
            } catch (e: Exception) {
                Log.w(TAG, "sync failed: ${e.message}")
            } finally {
                isSyncing = false
            }
        }
    }

    private suspend fun pushLocalLogs() {
        val logs = nutritionManager.logs
        if (logs.isEmpty()) return

        val logsArray = buildJsonArray {
            for (log in logs) {
                add(buildJsonObject {
                    put("clientId", log.id)
                    put("dishName", log.food)
                    put("calories", log.calories)
                    put("proteinG", log.protein)
                    put("carbsG", log.carbs)
                    put("fatsG", log.fats)
                    if (log.micros != null) {
                        put("micronutrients", buildJsonObject {
                            for ((k, v) in log.micros) put(k, v)
                        })
                    }
                    if (log.recipe != null) put("healthierRecipe", log.recipe)
                    put("loggedAt", epochMillisToIso(log.timeEpochMillis))
                })
            }
        }

        val body = buildJsonObject { put("logs", logsArray) }
        networkService.post("/api/sync/push", body)
    }

    private suspend fun pullRemoteLogs() {
        val path = if (lastSyncTimeMillis > 0) {
            "/api/sync/pull?since=${epochMillisToIso(lastSyncTimeMillis)}"
        } else {
            "/api/sync/pull"
        }

        val responseText = networkService.get(path)
        val response = json.decodeFromString(SyncPullResponse.serializer(), responseText)
        if (response.data.isEmpty()) return

        val existingIds = nutritionManager.logs.map { it.id }.toSet()
        val newLogs = response.data
            .filter { it.id !in existingIds }
            .map { remote ->
                FoodLog(
                    id = remote.id,
                    food = remote.dishName,
                    calories = remote.calories?.toInt() ?: 0,
                    protein = remote.proteinG?.toInt() ?: 0,
                    carbs = remote.carbsG?.toInt() ?: 0,
                    fats = remote.fatsG?.toInt() ?: 0,
                    micros = remote.micronutrients,
                    recipe = remote.healthierRecipe,
                    timeEpochMillis = isoToEpochMillis(remote.loggedAt),
                )
            }

        if (newLogs.isNotEmpty()) {
            nutritionManager.addSyncedLogs(newLogs)
        }
    }

    private fun epochMillisToIso(millis: Long): String =
        DateTimeFormatter.ISO_INSTANT.format(Instant.ofEpochMilli(millis))

    private fun isoToEpochMillis(iso: String): Long =
        try { Instant.parse(iso).toEpochMilli() } catch (_: Exception) { System.currentTimeMillis() }

    companion object {
        private const val TAG = "SyncManager"
    }
}

@Serializable
private data class SyncPullResponse(
    val success: Boolean,
    val data: List<RemoteFoodLog>,
)

@Serializable
private data class RemoteFoodLog(
    val id: String,
    val dishName: String,
    val calories: Double? = null,
    val proteinG: Double? = null,
    val carbsG: Double? = null,
    val fatsG: Double? = null,
    val micronutrients: Map<String, String>? = null,
    val healthierRecipe: String? = null,
    val loggedAt: String,
)
